yb-voyager/src/query/queryissue/detectors.go

```diff
@@ -187,3 +187,69 @@ func (d *RangeTableFuncDetector) GetIssues() []QueryIssue {
 	}
 	return issues
 }
+
+type CopyCommandUnsupportedConstructsDetector struct {
+	query                          string
+	copyFromWhereConstructDetected bool
+	copyOnErrorConstructDetected   bool
+}
+
+func NewCopyCommandUnsupportedConstructsDetector(query string) *CopyCommandUnsupportedConstructsDetector {
+	return &CopyCommandUnsupportedConstructsDetector{
+		query: query,
+	}
+}
+
+// Detect if COPY command uses unsupported syntax i.e. COPY FROM ... WHERE and COPY... ON_ERROR
+func (d *CopyCommandUnsupportedConstructsDetector) Detect(msg protoreflect.Message) error {
+	// Check if the message is a COPY statement
+	if msg.Descriptor().FullName() != queryparser.PG_QUERY_COPYSTSMT_NODE {
+		return nil // Not a COPY statement, nothing to detect
+	}
+
+	// Check for COPY FROM ... WHERE clause
```

Comment:
Simplified code from what I was discussing in the morning: 
I tried running your tests and they all passed, but do take a deeper look! 
cc: @sanyamsinghal @priyanshi-yb 
```
	protoMsg := msg.Interface().(proto.Message)
	copyStmt := protoMsg.(*pg_query.CopyStmt)

	if copyStmt.IsFrom && copyStmt.WhereClause != nil {
		d.copyFromWhereConstructDetected = true
	}
	for _, option := range copyStmt.Options {
		defElemNode := option.GetDefElem()
		if defElemNode != nil {
			if defElemNode.Defname == "on_error" {
				d.copyOnErrorConstructDetected = true
				break
			}
		}
	}

```

Reply:
I think yes this makes sense. It reduces the number of lines and readable.

Reply:
Lets stick to protomsg for this codepath, reason being the required info is not directly availabe in the struct and extraction of defElem can be generic and reused multiple times.

It can be simplified like this using existing helper functions:

```
FromField := GetMessageField(msg, "is_from")
whereField := GetMessageField(msg, "where_clause")
if FromField != nil && whereField != nil {
     d.copyFromWhereConstructDetected = true
}

defNames := TraverseAndExtractDefNamesFromDefElem(msg)
if slices.Contains(defNames, "on_error") {
    d.copyOnErrorConstructDetected = true
}

```
Refer my [PR](https://github.com/yugabyte/yb-voyager/pull/2108/files#:~:text=func%20TraverseAndExtractDefNamesFromDefElem(msg%20protoreflect.Message)%20%5B%5Dstring) for `TraverseAndExtractDefNamesFromDefElem`

Reply:
just to be on the same page, there are two different issues here. 

For the first one, isFrom and whereClause are directly available in `CopyStmt` struct, so would prefer dealing with the struct directly for that issue. But `GetMessageField(msg, "is_from"` isn't too bad either from a readability standpoint. I'll leave it up to you.

For the DefElem one, we can use the new function that Sanyam is introducing, but then you'll have to wait until that PR is merged.

Reply:
I had to add a new function GetBoolField() for the is_from clause. I have used the GetMessageField and GetBoolField approach rn for the first one.
For the second one, I have used Sanyam's function in my PR.